validator/src/metrics/mod.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------------
+
+use std::collections::HashMap;
+
+use cpython::{NoArgs, ObjectProtocol, PyDict, PyModule, PyObject, Python, ToPyObject};
+
+pub fn get_collector<S: AsRef<str>>(name: S) -> MetricsCollectorHandle {
+    let gil = Python::acquire_gil();
+    let py = gil.python();
+    let py_metrics = py.import("sawtooth_validator.metrics").unwrap();
+    let py_collector = py_metrics
+        .call(py, "get_collector", (name.as_ref(),), None)
+        .unwrap();
+    MetricsCollectorHandle {
+        py_collector,
+        py_metrics,
+    }
+}
+
+pub enum Level {
+    Debug,
+    Info,
+    Perf,
+}
+
+impl Default for Level {
+    fn default() -> Self {
+        Level::Info
+    }
+}
+
+fn into_level_str(level: Level) -> &'static str {
+    use self::Level::*;
+    match level {
+        Debug => "DEBUG",
+        Info => "INFO",
+        Perf => "PERF",
+    }
+}
+
+pub struct MetricsCollectorHandle {
+    py_collector: PyObject,
+    py_metrics: PyModule,
+}
+
+impl MetricsCollectorHandle {
+    pub fn counter<S: AsRef<str>>(
+        &self,
+        metric_name: S,
+        level: Option<Level>,
+        tags: Option<HashMap<String, String>>,
+    ) -> Counter {
+        Counter {
+            py_counter: self._create_metric("counter", metric_name, level, tags),
+        }
+    }
+
+    pub fn gauge<S: AsRef<str>>(
+        &self,
+        metric_name: S,
+        level: Option<Level>,
+        tags: Option<HashMap<String, String>>,
+    ) -> Gauge {
+        Gauge {
+            py_gauge: self._create_metric("gauge", metric_name, level, tags),
+        }
+    }
+
+    pub fn timer<S: AsRef<str>>(
+        &self,
+        metric_name: S,
+        level: Option<Level>,
+        tags: Option<HashMap<String, String>>,
+    ) -> Timer {
+        Timer {
+            py_timer: self._create_metric("timer", metric_name, level, tags),
+        }
+    }
+
+    fn _create_metric<S: AsRef<str>>(
```

Comment:
I don't think the leading underscore is needed, since the method is enforced as private.

Reply:
old habits die hard